source/server/http/admin.cc

```diff
@@ -529,17 +544,62 @@ Http::Code AdminImpl::handlerClusters(absl::string_view url, Http::HeaderMap& re
   return Http::Code::OK;
 }
 
-// TODO(jsedgwick) Use query params to list available dumps, selectively dump, etc
-Http::Code AdminImpl::handlerConfigDump(absl::string_view, Http::HeaderMap& response_headers,
+Http::Code AdminImpl::handlerConfigDump(absl::string_view url, Http::HeaderMap& response_headers,
                                         Buffer::Instance& response, AdminStream&) const {
+  Http::Utility::QueryParams query_params = Http::Utility::parseQueryString(url);
+  const auto resource = resourceParam(query_params);
+  const auto mask = maskParam(query_params);
+
   envoy::admin::v2alpha::ConfigDump dump;
+  bool resource_found = false;
   for (const auto& key_callback_pair : config_tracker_.getCallbacksMap()) {
     ProtobufTypes::MessagePtr message = key_callback_pair.second();
     RELEASE_ASSERT(message, "");
+
+    Protobuf::FieldMask* field_mask = new Protobuf::FieldMask;
+    if (mask.has_value()) {
+      ProtobufUtil::FieldMaskUtil::FromString(mask.value(), field_mask);
+    }
+
+    if (resource.has_value()) {
+      auto field_descriptor = message.get()->GetDescriptor()->FindFieldByName(resource.value());
+      const Protobuf::Reflection* reflection = message.get()->GetReflection();
+      if (!field_descriptor) {
+        continue;
+      } else if (!field_descriptor->is_repeated()) {
+        response.add(fmt::format("{} is not a repeated field. Use ?mask={} to get only this field",
+                                 field_descriptor->name(), field_descriptor->name()));
+        return Http::Code::BadRequest;
+      }
+      resource_found = true;
+
+      auto repeated =
+          reflection->MutableRepeatedPtrField<Protobuf::Message>(message.get(), field_descriptor);
```

Comment:
I don't believe that it modifies the config map as each callback returns a new unique ptr to a newly created and populated *ConfigDump message, e.g. https://github.com/envoyproxy/envoy/blob/master/source/server/listener_manager_impl.cc#L231

I started to write a test, but I don't think it ended up testing much because it is dependent on the implementation of the callback (like the one linked above) and the callbacks are re-implemented in the tests.